controllers/runnerpool_controller.go

```diff
@@ -0,0 +1,229 @@
+				r.Log.Error(err, "failed to remove finalizer", "name", req.NamespacedName)
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("removed finalizer", "name", req.NamespacedName)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, d, func() error {
+		return r.updateDeploymentWithRunnerPool(rp, d)
+	})
+	if err != nil {
+		r.Log.Error(err, "unable to create-or-update Deployment", "name", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("completed create-or-update successfully with status "+string(op), "name", req.NamespacedName)
+
+	rp.Status.Bound = true
+	err = r.Status().Update(ctx, rp)
+	if err != nil {
+		r.Log.Error(err, "failed to update status", "name", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *RunnerPoolReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&actionsv1alpha1.RunnerPool{}).
+		Owns(&appsv1.Deployment{}).
+		Complete(r)
+}
+
+func (r *RunnerPoolReconciler) addFinalizer(ctx context.Context, rp *actionsv1alpha1.RunnerPool) error {
+	rp2 := rp.DeepCopy()
+	controllerutil.AddFinalizer(rp2, runnerPoolFinalizer)
+	patch := client.MergeFrom(rp)
+	return r.Patch(ctx, rp2, patch)
+}
+
+func (r *RunnerPoolReconciler) removeFinalizer(ctx context.Context, rp *actionsv1alpha1.RunnerPool) error {
+	rp2 := rp.DeepCopy()
+	controllerutil.RemoveFinalizer(rp2, runnerPoolFinalizer)
+	patch := client.MergeFrom(rp)
+	return r.Patch(ctx, rp2, patch)
+}
+
+func (r *RunnerPoolReconciler) cleanUpOwnedResources(ctx context.Context, namespacedName types.NamespacedName) error {
+	d := &appsv1.Deployment{}
+	err := r.Get(ctx, namespacedName, d)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	return r.Delete(ctx, d)
+}
+
+func (r *RunnerPoolReconciler) updateDeploymentWithRunnerPool(rp *actionsv1alpha1.RunnerPool, d *appsv1.Deployment) error {
+	rp2 := rp.DeepCopy()
+
+	depLabels := rp2.GetLabels()
+	if depLabels == nil {
+		depLabels = make(map[string]string)
+	}
+	depLabels[constants.RunnerOrgLabelKey] = r.organizationName
+	depLabels[constants.RunnerRepoLabelKey] = rp.Spec.RepositoryName
+	d.ObjectMeta.Labels = depLabels
+
+	d.Spec.Replicas = rp2.Spec.Replicas
+	d.Spec.Selector = rp2.Spec.Selector
+	d.Spec.Strategy = rp2.Spec.Strategy
+
+	podLabels := rp2.Spec.Template.ObjectMeta.Labels
+	if podLabels == nil {
+		podLabels = make(map[string]string)
+	}
+	podLabels[constants.RunnerOrgLabelKey] = r.organizationName
+	podLabels[constants.RunnerRepoLabelKey] = rp.Spec.RepositoryName
+	d.Spec.Template.ObjectMeta.Labels = podLabels
+
+	d.Spec.Template.ObjectMeta.Annotations = rp2.Spec.Template.ObjectMeta.Annotations
+	d.Spec.Template.Spec = rp2.Spec.Template.Spec
```

Comment:
`deployment.spec.template` contains a lot of default values.
I think a simple overwrite will lead to unnecessary updates.

Please refer moco's implementation:
https://github.com/cybozu-go/moco/blob/e0cb68e1b1432973249367e9806ea8587c3ce4f1/controllers/mysqlcluster_controller.go#L800

Reply:
That really helps. Thank you.